File: contents/forward_euler_method/code/java/ForwardEuler.java
public class ForwardEuler {
    private static double[] solveEuler(double timestep, int n) {
        double[] eulerResult = new double[n];

        //Setting the initial condition
        eulerResult[0] = 1;
        for(int i = 1; i < eulerResult.length; i++) {
            eulerResult[i] = eulerResult[i - 1] - (3 * eulerResult[i - 1] * timestep);
        }
        return eulerResult;
    }

    private static boolean checkResult(double[] eulerResult, double threshold, double timestep) {
        boolean isApprox = true;

        for(int i = 0; i < eulerResult.length; i++) {
            double time = i * timestep;
            double solution = Math.exp(-3 * time);
            if(Math.abs(eulerResult[i] - solution) > threshold) {
                System.out.println(eulerResult[i] + " " + solution);
                isApprox = false;
            }
        }

        return isApprox;
    }

    public static void main(String[] args) {
        double timestep = 0.1;
        int n = 100;
        double threshold = 0.1;

        double[] eulerResult = solveEuler(timestep, n);
        boolean isApprox = checkResult(eulerResult, threshold, timestep);

        System.out.println(isApprox);
    }
}

File: contents/forward_euler_method/code/nim/forwardeuler.nim
import math

proc solveEuler(timestep: float, n: int): seq[float] =
  var res = newSeq[float](n)
  
  res[0] = 1.0

  for i in 1 .. n - 1:
    res[i] = res[i - 1] - 3 * res[i - 1] * timestep

  return res

proc check(res: seq[float], timestep, threshold: float): bool =
  var approx: bool = true;

  for i in 0 .. len(res) - 1:
    let solution: float = exp(-3.0 * float(i) * timestep)
    if abs(res[i]) - solution > threshold:
      echo res[i]
      echo solution
      approx = false

  return approx

const
  timestep: float = 0.1
  n: int = 100
  threshold: float = 0.1
  eulerResult: seq[float] = solveEuler(timestep, n)
  approx: bool = check(eulerResult, threshold, timestep)

echo approx
